sql/src/main/java/org/apache/druid/sql/calcite/external/TableAppendMacro.java

```diff
@@ -0,0 +1,344 @@
+    {
+      return table.unwrapOrThrow(DatasourceTable.class);
+    }
+
+    public RowSignature getRowSignature()
+    {
+      return getDataSourceTable().getRowSignature();
+    }
+
+    public DataSource getDataSource()
+    {
+      return getDataSourceTable().getDataSource();
+    }
+  }
+
+  private List<TableOperand> getTables(SqlCallBinding callBinding, SqlValidatorCatalogReader catalogReader)
+  {
+    List<TableOperand> tables = new ArrayList<>();
+    for (int i = 0; i < callBinding.getOperandCount(); i++) {
+      if (!callBinding.isOperandLiteral(i, false)) {
+        throw DruidSqlValidator.buildCalciteContextException(
+            "All arguments of call to macro "
+                + "APPEND should be literal. Actual argument #"
+                + i + " is not literal",
+                callBinding.operand(i)
+        );
+      }
+      @Nullable
+      String tableName = callBinding.getOperandLiteralValue(i, String.class);
+      ImmutableList<String> tableNameList = ImmutableList.<String>builder().add(tableName).build();
+      SqlValidatorTable table = catalogReader.getTable(tableNameList);
+      if (table == null) {
+        throw DruidSqlValidator.buildCalciteContextException(
+            StringUtils.format("Table [%s] not found", tableName),
+            callBinding.operand(i)
+        );
+      }
+      tables.add(new TableOperand(callBinding.operand(i), table));
+    }
+    return tables;
+  }
+
+  static class AppendDatasourceMetadata implements DatasourceMetadata
+  {
+    private final RowSignature values;
+    private final DataSource dataSource;
+
+    public AppendDatasourceMetadata(RowSignature values, List<DataSource> dataSources)
+    {
+      this.values = values;
+      this.dataSource = new UnionDataSource(dataSources);
+    }
+
+    @Override
+    public boolean isJoinable()
+    {
+      return false;
+    }
+
+    @Override
+    public boolean isBroadcast()
+    {
+      return false;
+    }
+
+    @Override
+    public DataSource dataSource()
+    {
+      return dataSource;
+    }
+  }
+
+  private AppendDatasourceMetadata buildUnionDataSource(List<TableOperand> tables)
+  {
+    List<DataSource> dataSources = new ArrayList<>();
+    Map<String, ColumnType> fields = new LinkedHashMap<>();
+    Builder rowSignatureBuilder = RowSignature.builder();
+    for (TableOperand table : tables) {
+      RowSignature rowSignature = table.getRowSignature();
+      for (String columnName : rowSignature.getColumnNames()) {
+        ColumnType currentType = rowSignature.getColumnType(columnName).get();
+        ColumnType existingType = fields.get(columnName);
+
+        if (existingType == null || existingType.equals(currentType)) {
+          fields.put(columnName, currentType);
+        } else {
+          try {
+            ColumnType commonType = ColumnType.leastRestrictiveType(currentType, existingType);
+            fields.put(columnName, commonType);
+          }
+          catch (Exception e) {
```

Comment:
I think we should only catch `Types.IncompatibleTypeException` - is there a reason for a catch all?